Feature/hd 4220 last block size

fferrandis opened the pull request:
This pull request introduces significant improvements and refactoring to the buffer matrix management for erasure coding, focusing on more efficient handling of data fragments, especially for cases where the last subgroup (stripe) is not a full block. The changes include a new buffer format, more flexible offset calculations, improved test coverage, and a new C header for backend interoperability.

**Key changes:**

### Buffer Matrix Refactoring and New Format Support

* Introduced a new `BufferInfo` struct to encapsulate buffer metadata, separated from the main `BufferMatrix` data structure, and refactored related logic for clarity and maintainability. Added methods to support both old and new buffer layouts, with explicit `UseNewFormat` and `UseOldFormat` setters and corresponding offset calculation logic. (`buffer.go` [[1]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L8-R97) [[2]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R112-L92) [[3]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R165-R242) [[4]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L125-R251) [[5]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R299-R305) [[6]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R317-R341)

* Implemented logic to handle the last subgroup differently when the data size is not a multiple of the block size, ensuring correct data alignment and minimizing wasted space in the last fragments. (`buffer.go` [[1]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R165-R242) [[2]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R299-R305) [[3]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R317-R341)

### Expanded and Improved Testing

* Removed outdated or redundant tests and benchmarks, and added comprehensive new tests to cover the new buffer format, offset calculations, subgroup size computations, and block mapping. This ensures correctness of the new logic and robust coverage for edge cases. (`buffer_test.go` [[1]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L74-L116) [[2]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L134-R93) [[3]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L163-R281)

### Backend Interoperability

* Added a new C header file `backend.h` defining the `encode_chunk_context` structure and a suite of utility functions to facilitate interaction between Go and C code for erasure coding backends. (`backend.h` [backend.hR1-R70](diffhunk://#diff-f142809213f44ad3f9647b0664a4272e61cd1096c051006621c94cc1c628ea47R1-R70))

### Minor Improvements

* Added a `// nolint:gosec` comment to suppress a linter warning for pointer arithmetic in `memalign.go`. (`memalign.go` [memalign.goR9](diffhunk://#diff-97c9d06b50cbe1f96309c4f4a8ada847ea6a372632944fc37a8eb74a49b2e574R9))

These changes collectively modernize the buffer management for erasure coding, improve performance for non-aligned data sizes, and lay groundwork for further backend integration.

**References:** [[1]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L8-R97) [[2]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R112-L92) [[3]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R165-R242) [[4]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L125-R251) [[5]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R299-R305) [[6]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R317-R341) [[7]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L74-L116) [[8]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L134-R93) [[9]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L163-R281) [[10]](diffhunk://#diff-f142809213f44ad3f9647b0664a4272e61cd1096c051006621c94cc1c628ea47R1-R70) [[11]](diffhunk://#diff-97c9d06b50cbe1f96309c4f4a8ada847ea6a372632944fc37a8eb74a49b2e574R9)

fferrandis committed:
HD-4220: Implement last block size optimisation

- separate C implementations in dedicated files
- improve encoding and decoding functionality
- add new tests
- improvements to the encoding and decoding processes

bert-e commented:
# Hello fferrandis,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/erasurecode/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

fferrandis committed:
HD-4220: Enable new style by default

- enable new buffer style
- rework all tests to only use *Matrix

fferrandis committed:
HD-4220: adapt get range

fferrandis committed:
HD-4220: linearize and reconstruct testing

fferrandis committed:
HD-4220: Last block size

Store in EncodeData structure the realsize
With old format, it contains the FragLen itself
With new format, this is sticked to the real data.

In DecodeMatrix, we also manage the last stripe size to ensure we read the
header.ChunkSize value instead of reading piecesize bytes

fferrandis commented on backend.c:
```diff
@@ -0,0 +1,408 @@
+#include <liberasurecode/erasurecode.h>
```
this is only a move from code that was directly inline in GO file to a proper dedicated C file.

fferrandis commented on backend.go:
```diff
@@ -591,25 +185,13 @@ func InitBackend(params Params) (Backend, error) {
 			max: params.MaxBlockSize,
 		}
 	}
 
```
as we dont use jerasure, this is removed. (this new feature can't work with jerasure anyway)

fferrandis commented on backend.go:
```diff
@@ -684,7 +274,15 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
 
 	for i := 0; i < int(ctx.number_of_subgroup); i++ {
 		go func(nth int) {
```
this is the main diff of the encode part. We make a distinction when it comes to the last lines/stripes. All the stripes are supposed to be "chunksize" long, but not the last one because of the new implementation

fferrandis commented on backend.go:
```diff
@@ -697,7 +295,7 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
 				C.my_liberasurecode_encode_buffermatrix_cleanup(
 					backend.libecDesc, C.size_t(ctx.frags_len), ctx.datas, ctx.codings)
-			}},
+			}, totLen},
```
we store the real len of the slice. We could also reslice it to this len, but it was easier from the debug point of view to keep the slice with the same slice regardless of the usage of the new format, and store the real len alongside.

fferrandis commented on backend.go:
```diff
@@ -716,57 +314,7 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
-	}}, nil
-}
-
-// EncodeMatrix encodes data in small subpart of chunkSize bytes
```
EncodeMatrix isn't used in hdcontroller. So all the tests using it has been updated to replace the usage of EncodeMatrix by EncodeMatrixWithBufferMatrix

fferrandis commented on backend.go:
```diff
@@ -972,8 +519,9 @@ func (backend *Backend) DecodeMatrix(frags []ValidatedFragment, pieceSize int) (
 	var totLen int64
 	for i := 0; i < chunkInfo.NrChunk; i++ {
 		vect := make([][]byte, len(frags))
```
when decoding stripes, we can't assume that all stripes are "chunksized long". When using the new format, the last stripe of a fragment is often smaller (that is the prupose of this PR and this new format). The nextbound is computed accordingly.

fferrandis commented on backend.go:
```diff
@@ -1050,41 +598,69 @@ type RangeMatrix struct {
  *     p4 [-[*]- -]
  *
  */
```
this is mainly variable renaming to help to understand what we are manipulating in the matrix. That why we have lines, celles and columns.

fferrandis commented on backend.go:
```diff
@@ -1050,41 +598,69 @@ type RangeMatrix struct {
+	totalLines := (maxDataSize + lineSize - 1) / lineSize
+	isLastStripe := (lineEnd == totalLines-1)
 
 	/* When wrapping around, we read the full groups. */
```
this is the main change of GetRangeMatrix : we are going to read a full group (ie a full line of stripes) in case of two condition : 
- first one already existed
- the second one is new and is true when data is in the last stripe line. 
As we cannot guess how the data is stored (using the old format or the new one) we chose to read all the stripes and then, based on the stripes header, get properly the data

fferrandis reviewed (commented):
some comments to help the review

fferrandis committed:
HD-4220: add accessor to encodedata struct

fferrandis committed:
HD-4220: ReconstructMatrix should size the generated buffer

liaud commented on backend.go:
```diff
@@ -637,13 +660,13 @@ func (backend *Backend) GetRangeMatrix(startIncl, endIncl, pieceSize, fragSize i
 	linearizedRangeStartIncl := startIncl - dataOffset
 
 	/* Decoding always works on a group boundary. */
```
You may want to update the comment to not mention "group" but "column" given the variable names.

benoit-a commented on backend_test.go:
```diff
@@ -1242,3 +1319,314 @@ func TestRangeMatrix(t *testing.T) {
+	bm := NewBufferMatrix(chunkSize, len(buf), backend.K)
+	if useNewFormat {
+		bm.UseNewFormat()
+	}
```
Since new format is default, should we consider adding an `else { /* use old format */ }`

fferrandis replied:
yes, nice catch

fferrandis committed:
HD-4220: test GetRangeHelper with old format

fferrandis commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **approve**

liaud commented on backend.c:
```diff
@@ -0,0 +1,408 @@
+    string_off += psize;
+  }
+  return dest;
+}
+
+bool check_matrix_fragment(char *frag, int frag_len, int piecesize) {
+  size_t offset = 0;
+
+  while (offset < frag_len) {
+    if (is_invalid_fragment_header((fragment_header_t *)&frag[offset])) {
+      return false;
+    }
+    offset += piecesize + getHeaderSize();
+  }
+
```
Maybe an alternative would be to check that the remaining has just enough data for the header before calling `is_invalid_fragment_header` in the loop above, then check that the header returned size valid w.r.t to the provided buffer.

liaud commented on backend.go:
```diff
@@ -684,7 +280,15 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
-			r := C.encode_chunk_buffermatrix(backend.libecDesc, pData, pDataLen, nbFrags, &ctx, C.int(nth))
+			fragLen := C.size_t(chunkSize)
+			// last subgroup has a different size
+			if i == int(ctx.number_of_subgroup)-1 {
```
This suggestion looks correct. I guess it works by "chance" because the last goroutine will have the correct shared `i`.

liaud commented on backend_test.go:
```diff
@@ -813,26 +793,147 @@ func TestEncodeM(t *testing.T) {
+				frags2 = append(frags2, encoded.Data[i][rangeM.InFragRangeStartIncl:rangeM.InFragRangeEndExcl])
+			}
+
+			// now do the same test, but this time, insted of linearizing, we are going to reconstruct the stripes
```
```suggestion
			// now do the same test, but this time, instead of linearizing, we are going to reconstruct the stripes
```

fferrandis committed:
HD-4220: fix concurrency potential bug

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.0`




Please check the status of the associated issue HD-4220.

Goodbye fferrandis.




*The following options are set:* **approve**
